File: week05/3/prob2/delete_duplicates.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def deleteDuplicates(self, head):
    
    # empty or one node
    if not head or not head.next:
        return head

    # initialixation: traversal pointer
    curr = head

    # traversal
    while curr and curr.next:
        if curr.val == curr.next.val:
            # bypass next node
            curr.next = curr.next.next
        else:
            # move pointer to next node
            curr = curr.next

    return head

File: week05/3/prob2/delete_duplicates.c
#include <stdlib.h>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* deleteDuplicates(struct ListNode* head) {
    // empty or one node
    if (head == NULL || head->next == NULL) {
        return head;
    }

    // initialization: traversal pointer
    struct ListNode* curr = head;

    // traversal
    while (curr != NULL && curr->next != NULL) {
        if (curr->val == curr->next->val) {
            // bypass next node
            struct ListNode* temp = curr->next; // temporary node to hold the duplicate
            curr->next = curr->next->next; // bypass duplicate node
            free(temp); // free memory of duplicate node
        } else {
            // move pointer to next node
            curr = curr->next;
        }
    }

    return head;
}